pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/relocation/RealtimeSegmentRelocator.java

```diff
@@ -0,0 +1,278 @@
+      }
+    });
+  }
+
+  public void start() {
+    LOGGER.info("Starting realtime segment relocator");
+
+    _executorService.scheduleWithFixedDelay(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          runRelocation();
+        } catch (Exception e) {
+          LOGGER.warn("Caught exception while running realtime segment relocator", e);
+        }
+      }
+    }, 120, _runFrequencySeconds, TimeUnit.SECONDS);
+  }
+
+  public void stop() {
+    _executorService.shutdown();
+  }
+
+  /**
+   * Check all tables. Perform relocation of segments if table is realtime and relocation is required
+   */
+  public void runRelocation() {
+    if (!_pinotHelixResourceManager.isLeader()) {
+      LOGGER.info("Skipping realtime segment relocation, not leader!");
+      return;
+    }
+
+    LOGGER.info("Starting relocation of realtime segments");
+    List<String> allRealtimeTableNames = _pinotHelixResourceManager.getAllRealtimeTables();
+
+    for (final String tableNameWithType : allRealtimeTableNames) {
+      try {
+        LOGGER.info("Starting relocation of segments for table: {}", tableNameWithType);
+
+        TableConfig tableConfig = _pinotHelixResourceManager.getRealtimeTableConfig(tableNameWithType);
+        final RealtimeTagConfig realtimeTagConfig = new RealtimeTagConfig(tableConfig, _helixManager);
+        if (!realtimeTagConfig.isRelocateCompletedSegments()) {
+          LOGGER.info("Skipping relocation of segments for {}", tableNameWithType);
+          return;
+        }
+
+        Function<IdealState, IdealState> updater = new Function<IdealState, IdealState>() {
+          @Nullable
+          @Override
+          public IdealState apply(@Nullable IdealState idealState) {
+            if (!idealState.isEnabled()) {
+              LOGGER.info("Skipping relocation of segments for {} since ideal state is disabled", tableNameWithType);
+              return null;
+            }
+            relocateSegments(realtimeTagConfig, idealState);
+            return idealState;
+          }
+        };
+
+        HelixHelper.updateIdealState(_helixManager, tableNameWithType, updater,
+            RetryPolicies.exponentialBackoffRetryPolicy(5, 1000, 2.0f));
+      } catch (Exception e) {
+        LOGGER.error("Exception in relocating realtime segments of table {}", tableNameWithType, e);
+      }
+    }
+    LOGGER.info("Realtime segment relocation completed");
+  }
+
+  /**
+   * Given a realtime tag config and an ideal state, relocate the segments
+   * which are completed but still hanging around on consuming servers, one replica at a time
+   * @param  realtimeTagConfig
+   * @param idealState
+   */
+  protected void relocateSegments(RealtimeTagConfig realtimeTagConfig, IdealState idealState) {
+
+    List<String> consumingServers = _helixAdmin.getInstancesInClusterWithTag(_helixManager.getClusterName(),
+        realtimeTagConfig.getConsumingServerTag());
+    if (consumingServers.isEmpty()) {
```

Comment:
do we want to retry in the previous method if there is config error?

Reply:
Hmm, no point in retrying for config error. Will see if i can restructure it a bit